pytorch_tabnet/metrics.py

```diff
@@ -0,0 +1,325 @@
+from dataclasses import dataclass
+from typing import List
+import numpy as np
+from sklearn.metrics import (
+    roc_auc_score,
+    mean_squared_error,
+    mean_absolute_error,
+    accuracy_score,
+    log_loss,
+    balanced_accuracy_score,
+)
+
+
+@dataclass
+class MetricContainer:
+    """Container holding a list of metrics.
+
+    Parameters
+    ----------
+    metric_names : list of str
+        List of metric names.
+    prefix : str
+        Prefix of metric names.
+
+    """
+
+    metric_names: List[str]
+    task: str
+    prefix: str = ""
+
+    def __post_init__(self):
+        self.metrics = Metric.get_metrics_by_names(self.metric_names, self.task)
+        self.names = [self.prefix + name for name in self.metric_names]
+
+    def __call__(self, y_true, y_pred):
+        """Compute all metrics and store into a dict.
+
+        Parameters
+        ----------
+        y_true: np.ndarray
+            Target matrix or vector
+        y_pred: np.ndarray
+            Score matrix or vector
+
+        Returns
+        -------
+        dict
+            Dict of metrics ({metric_name: metric_value}).
+
+        """
+        logs = {}
+        for metric in self.metrics:
+            if isinstance(y_pred, list):
+                res = np.mean(
+                    [metric(y_true[:, i], y_pred[i]) for i in range(len(y_pred))]
+                )
+            else:
+                res = metric(y_true, y_pred)
+            logs[self.prefix + metric._name] = res
+        return logs
+
+
+class Metric:
+    def __call__(self, y_true, y_pred):
+        raise NotImplementedError("Custom Metrics must implement this function")
+
+    @classmethod
+    def get_metrics_by_names(cls, names, task):
+        """Get list of metric classes.
+
+        Parameters
+        ----------
+        cls : Metric
+            Metric class.
+        names : list
+            List of metric names.
+        task : str.
+            'classification' or 'regression'.
+
+        Returns
+        -------
+        metrics : list
+            List of metric classes.
+
+        """
+        available_metrics = cls.__subclasses__()
+        available_names = [metric()._name for metric in available_metrics]
+        metrics = []
+        for name in names:
+            assert name in available_names, f"{name} is not available"
+            idx = available_names.index(name)
+            metric = available_metrics[idx]()
+            assert metric._task == task, f"{name} is not compatible for {task}"
+            metrics.append(metric)
+        return metrics
+
+
+class AUC(Metric):
+    """
+    AUC.
+    """
+
+    def __init__(self):
+        self._name = "auc"
+        self._maximize = True
+        self._task = 'classification'
```

Comment:
`self._task` should not be mandatory when creating a custom metric. Should we simple remove it?
I think default metrics given should be understood well enough by users so we don't need to check if it's ok to use it.
Errors will be at user expense.

Reply:
I deleted it